### PR TITLE
http conn man: allow to upper-bound downstream connection lifetime.

### LINKED ISSUE #8302: http: Allow upper bounding lifetime of downstream connections
*Description*:
We use envoy as user-facing edge proxy behind lvs/ipvs.
During the restart of the envoy proxy connections tend to migrate to other processes.
This leads to cascading load imbalance (the last process to restart has most connections, typically several times more than in steady mode).

Does it make sense to add configuration option to HTTP Connection Manager to gracefully drain a connection after certain timeout since connection establishment? (Similar to idle timeout)

### PR DESCRIPTION
**Pull request by veshij**
Description: Allow upper bounding lifetime of downstream connections 
Risk Level: Medium
Testing: unit test, integration test
Docs Changes: None
Release Notes: updated version_history.rst
Fixes #8302

### CODE REVIEW

**Comment by jmarantz** [comment]
please merge master.

File: api/envoy/api/v2/core/protocol.proto [hunk 1]
```diff
@@ -27,6 +27,14 @@ message HttpProtocolOptions {
   // Note that request based timeouts mean that HTTP/2 PINGs will not keep the connection alive.
   google.protobuf.Duration idle_timeout = 1;
 
+  // The maximum lifetime of downstream connection. The lifetime timeout is defined
+  // as a period since downstream connection was established.
+  // If not set, there is no lifetime timeout. When lifetime_timeout is reached the connection will
```
Comment (htuch) [inline api/envoy/api/v2/core/protocol.proto]: Nit: `lifetime_timeout`

File: api/envoy/api/v2/core/protocol.proto [hunk 2]
```diff
@@ -27,6 +27,14 @@ message HttpProtocolOptions {
   // Note that request based timeouts mean that HTTP/2 PINGs will not keep the connection alive.
   google.protobuf.Duration idle_timeout = 1;
 
+  // The maximum lifetime of downstream connection. The lifetime timeout is defined
```
Comment (htuch) [inline api/envoy/api/v2/core/protocol.proto]: I think this should just be "of a connection", and you can then qualify that it only works for downstream. There's no reason this can't be implemented to apply to upstream when these protocol options are used by cluster manager.

File: api/envoy/api/v2/core/protocol.proto [hunk 3]
```diff
@@ -27,6 +27,14 @@ message HttpProtocolOptions {
...
+  // as a period since downstream connection was established.
+  // If not set, there is no lifetime timeout. When lifetime_timeout is reached the connection will
+  // be closed. If the connection is an HTTP/2 connection a drain sequence will occur prior to
+  // closing the connection. See :ref:`drain_timeout
+  // <envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.drain_timeout>`.
+  google.protobuf.Duration lifetime_timeout = 3;
```
Comment (htuch) [inline api/envoy/api/v2/core/protocol.proto]: Nit: probably bike shedding, but I think I prefer `connection_maximum_duration` as more descriptive.

Comment (veshij) [reply api/envoy/api/v2/core/protocol.proto]: renamed to `max_connection_duration`

File: api/envoy/api/v2/core/protocol.proto (hunk 2 above)
Comment (veshij) [reply api/envoy/api/v2/core/protocol.proto]: Updated comment. What would be the best way add a note that it's implemented for downstream connections only?

File: source/common/http/conn_manager_impl.cc [hunk 4]
```diff
@@ -404,6 +415,19 @@ void ConnectionManagerImpl::onIdleTimeout() {
   }
 }
 
+void ConnectionManagerImpl::onConnectionDurationTimeout() {
+  ENVOY_CONN_LOG(debug, "max connection duration reached", read_callbacks_->connection());
+  stats_.named_.downstream_cx_max_duration_reached_.inc();
+  if (!codec_) {
+    // TODO(oleg): figure out if following is true for closing non-idle connections.
```
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.cc]: I believe it's fine to do a flush write.  In fact it might be useful to have an onTimeout utility function which can be called from onConectionDurationTimeout and onIdletimeout

File: source/common/http/conn_manager_impl.h [hunk 5]
```diff
@@ -675,6 +676,7 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
   // connection. When there are active streams it is disarmed in favor of each stream's
   // stream_idle_timer_.
   Event::TimerPtr connection_idle_timer_;
+  Event::TimerPtr connection_duration_timer_;
```
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.h]: mind adding comments?

File: api/envoy/api/v2/core/protocol.proto (hunk 2 above)
Comment (alyssawilk) [reply api/envoy/api/v2/core/protocol.proto]: Arguably an API shephard thing but I'd prefer we either have folks sign up for the complete functionality (if we think we want both) in which case a TODO and follow-up PR is OK, or we move this to the HCM with the understanding we're only doing it for downstream.

Comment (veshij) [reply api/envoy/api/v2/core/protocol.proto]: moved to HttpProtocolOptions

**Comment by veshij** [comment]
> Looks great overall! Can you please add an idle timeout integration test along with the unit test? There should be some in test/integration/idle_timeout_integration_test.cc you can crib off of fairly easily.

added an integration test

File: test/integration/connection_duration_ingegration_test.cc [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+          hcm.set_proxy_100_continue(true);
+        });
+    HttpProtocolIntegrationTest::initialize();
+  }
+
+  // TODO(htuch): This might require scaling for TSAN/ASAN/Valgrind/etc. Bump if
+  // this is the cause of flakes.
+  static constexpr uint64_t MaxConnectionDurationMs = 400;
+
+  bool enable_connection_duration_timeout_{false};
```
[the remark this answers is not in the document]
Comment (veshij) [reply test/integration/connection_duration_ingegration_test.cc]: cleaned up

File: test/integration/connection_duration_ingegration_test.cc [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+namespace {
+
+class ConnectionDurationIntegrationTest : public HttpProtocolIntegrationTest {
+public:
+  void initialize() override {
+    config_helper_.addConfigModifier(
+        [&](envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager& hcm)
+            -> void {
+          if (enable_connection_duration_timeout_) {
+            hcm.mutable_max_connection_duration()->set_seconds(0);
```
[the remark this answers is not in the document]
Comment (veshij) [reply test/integration/connection_duration_ingegration_test.cc]: cleaned up

File: test/integration/connection_duration_ingegration_test.cc [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            hcm.mutable_max_connection_duration()->set_nanos(MaxConnectionDurationMs * 1000 * 1000);
+          }
+
+          // For validating encode100ContinueHeaders() timer kick.
+          hcm.set_proxy_100_continue(true);
```
[the remark this answers is not in the document]
Comment (veshij) [reply test/integration/connection_duration_ingegration_test.cc]: cleaned up

File: test/integration/connection_duration_ingegration_test.cc [hunk 9]
```diff
@@ -0,0 +1,92 @@
+#include "test/integration/http_protocol_integration.h"
+#include "test/test_common/test_time.h"
+
+namespace Envoy {
+namespace {
+
+class ConnectionDurationIntegrationTest : public HttpProtocolIntegrationTest {
```
[the remark this answers is not in the document]
Comment (veshij) [reply test/integration/connection_duration_ingegration_test.cc]: Yep, now it looks better.
I was using idle_timeout_integration_test as a reference before.

File: api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto [hunk 10]
```diff
@@ -310,6 +310,13 @@ message HttpConnectionManager {
   // response is initiated. If not specified or set to 0, this timeout is disabled.
   google.protobuf.Duration request_timeout = 28;
 
+  // The maximum duration of a connection. The duration is defined as a period since a connection
+  // was established. If not set, there is no max duration. When max_connection_duration is reached
+  // the connection will be closed. If the connection is an HTTP/2 connection a drain sequence will
+  // occur prior to closing the connection. See :ref:`drain_timeout
+  // <envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.drain_timeout>`.
+  google.protobuf.Duration max_connection_duration = 36;
```
Comment (htuch) [inline api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto]: If we might want this one day on the upstream, I would prefer we move it to the protocol options. If we are certain it is only ever a downstream thing (which I would find surprising), then this is fine.

Comment (veshij) [reply api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto]: We might want to upper-bound upstream connections as well.
What would be the best way to mark is as 'not implemented' for upstream connections?

Comment (alyssawilk) [reply api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto]: My preference is if we're implementing it in one place to have it here, and if we move it that we implement in both places (which is a fairly big ask)
Happy to have @mattklein123 tie break if we prefer having it half-implemented in the right place, vs in the potentially wrong place but have config clarity :-)

Comment (mattklein123) [reply api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto]: My vote would be to move it into the common options if we think that it will eventually be implemented for upstream, mainly because the deprecation cycle is pretty painful.

Comment (htuch) [reply api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto]: I do share @alyssawilk 's concern over features being added to the API with no intention of Envoy-side support, but that's mostly focused on the situation where an API is featured with zero Envoy implementation. Having a partial implementation that delivers value, together with a TODO and tracking issue so that someone who cares about the behavior can follow-thru on later seems a lot more tenable to me.

Comment (veshij) [reply api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto]: I defined max_connection_duration in common options and added some stubs in upstream implementation.

File: test/integration/connection_duration_ingegration_test.cc [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  EXPECT_TRUE(response->complete());
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_total", 1);
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_200", 1);
+
+  ASSERT_TRUE(codec_client_->waitForDisconnect());
+  test_server_->waitForCounterGe("http.config_test.downstream_cx_max_duration_reached", 1);
+}
+
+// Test inflight request is processed correctly when timeout fires during request processing.
+TEST_P(ConnectionDurationIntegrationTest, InflightRequest) {
```
[the remark this answers is not in the document]
Comment (veshij) [reply test/integration/connection_duration_ingegration_test.cc]: hm, in some rare cases test finishes in some unreasonable time:
```
Protocols/ProtocolIntegrationTest.ConnDurationTimeoutBasic/IPv6_HttpDownstream_Http2Upstream (1179354 ms)
``` 
trying to figure out what's going on there.

[the remark this answers is not in the document]
Comment (veshij) [reply test/integration/connection_duration_ingegration_test.cc]: Fixed with explicitly setting drain_timeout in integration tests.

File: test/integration/protocol_integration_test.cc [hunk 12]
```diff
@@ -1355,16 +1363,20 @@ TEST_P(ProtocolIntegrationTest, ConnDurationInflightRequest) {
   test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_total", 1);
   test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_200", 1);
 
-  ASSERT_TRUE(codec_client_->waitForDisconnect());
+  ASSERT_TRUE(codec_client_->waitForDisconnect(std::chrono::milliseconds(10000)));
```
Comment (htuch) [inline test/integration/protocol_integration_test.cc]: This is probably OK, but if we see future flakes, this should be increase by some small fudge factor.

**Comment by veshij** [comment]
Is there any action items for me here?
I suppose some reviewers were added automatically when I accidentally pushed rebased branch.

File: include/envoy/upstream/upstream.h [hunk 13]
```diff
@@ -682,6 +682,11 @@ class ClusterInfo {
    */
   virtual const absl::optional<std::chrono::milliseconds> idleTimeout() const PURE;
 
+  /**
+   * @return optional maximum connection duration timeout for upstream connection pool connections.
+   */
+  virtual const absl::optional<std::chrono::milliseconds> maxConnectionDuration() const PURE;
```
Comment (mattklein123) [inline include/envoy/upstream/upstream.h]: If it's not implemented for upstream connections why was this added? Can it be removed?

File: api/envoy/api/v2/core/protocol.proto [hunk 14]
```diff
@@ -28,6 +28,14 @@ message HttpProtocolOptions {
   // If not specified, this defaults to 1 hour. To disable idle timeouts explicitly set this to 0.
   google.protobuf.Duration idle_timeout = 1;
 
+  // The maximum duration of a connection. The duration is defined as a period since a connection
+  // was established. If not set, there is no max duration. When max_connection_duration is reached
+  // the connection will be closed. If the connection is an HTTP/2 connection a drain sequence will
+  // occur prior to closing the connection. See :ref:`drain_timeout
+  // <envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.drain_timeout>`.
+  // Note: not implemented for upstream connections.
+  google.protobuf.Duration max_connection_duration = 3;
```
Comment (mattklein123) [inline api/envoy/api/v2/core/protocol.proto]: Please add a release note for this new feature with a ref link to the new configuration field.

File: api/envoy/api/v2/core/protocol.proto [hunk 15]
```diff
@@ -28,6 +28,14 @@ message HttpProtocolOptions {
   // If not specified, this defaults to 1 hour. To disable idle timeouts explicitly set this to 0.
   google.protobuf.Duration idle_timeout = 1;
 
+  // The maximum duration of a connection. The duration is defined as a period since a connection
+  // was established. If not set, there is no max duration. When max_connection_duration is reached
+  // the connection will be closed. If the connection is an HTTP/2 connection a drain sequence will
```
Comment (mattklein123) [inline api/envoy/api/v2/core/protocol.proto]: I don't think this is correct. From reading the code it looks like a drain sequence will happen for HTTP/1 as well?

File: source/common/upstream/upstream_impl.cc [hunk 16]
```diff
@@ -730,6 +730,13 @@ ClusterInfoImpl::ClusterInfoImpl(
     idle_timeout_ = std::chrono::hours(1);
   }
 
+  if (config.common_http_protocol_options().has_max_connection_duration()) {
```
Comment (mattklein123) [inline source/common/upstream/upstream_impl.cc]: Please revert all changes in this file. We can add this back later if/when implemented.

Comment (veshij) [reply source/common/upstream/upstream_impl.cc]: I was following suggestion about partial implementation in https://github.com/envoyproxy/envoy/pull/8591#discussion_r337322325.
I'm ok with either of approaches :)

Comment (mattklein123) [reply source/common/upstream/upstream_impl.cc]: I think by partial implementation @htuch meant downstream vs. upstream, not half of upstream. :) Thank you!

Comment (veshij) [reply source/common/upstream/upstream_impl.cc]: I think I just misread 'delivers partial value'. I'll revert these changes.

File: api/envoy/api/v2/core/protocol.proto (hunk 14 above)
Comment (veshij) [reply api/envoy/api/v2/core/protocol.proto]: done.

File: include/envoy/upstream/upstream.h (hunk 13 above)
Comment (veshij) [reply include/envoy/upstream/upstream.h]: removed.

File: api/envoy/api/v2/core/protocol.proto (hunk 15 above)
Comment (veshij) [reply api/envoy/api/v2/core/protocol.proto]: updated description

File: source/common/upstream/upstream_impl.cc (hunk 16 above)
Comment (veshij) [reply source/common/upstream/upstream_impl.cc]: reverted.
